man/hashid.7
.\" Manpage for hashid
.\" This file is part of hashID.
.\" 
.\" hashID is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\" 
.\" hashID is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with hashID.  If not, see <http://www.gnu.org/licenses/>.

.TH hashid 7 "28 January 2014" "v1.4" "hashid man page"
.SH NAME
\fBhashid\fP \- Identify the different types of hashes used to encrypt data

.SH SYNOPSIS
.br
.B hashid
[-h] [-a] [-m] [-j] [-o FILE] [--version] INPUT

.SH DESCRIPTION
.B hashID 
is a tool written in Python 3.x which supports the identification of over 205 unique hash types using regular expressions.

It is able to identify a single hash, parse a file or read files in a directory and identify the hashes within them.
.B hashID
is also capable of including the corresponding hashcat mode and/or JohnTheRipper format in its output.
Altough
.B hashID
is written in Python 3.x it should also work with Python 2.7.

.SH OPTIONS
.TP
\fB\-e, \-\-extended\fR
list all possible hash algorithms including salted passwords
.TP
\fB\-m, \-\-mode\fR
include corresponding hashcat mode in output
.TP
\fB\-j, \-\-john\fR
include corresponding JohnTheRipper format in output
.TP
\fB\-o FILE, \-\-outfile FILE\fR
write output to file (default: STDOUT)
.TP
\fB\-h, \-\-help\fR
show help message and exit
.TP
\fB\-\-version\fR
show program's version number and exit

.SH BUGS
If you find a bug, please report it at https://github.com/psypanda/hashID/issues

.SH AUTHOR
c0re (c0re@psypanda.org)

CHANGELOG
### Changelog

v2.9.0
- added support for piped input (thanks Max)
- added SHA1($salt.$pass.$salt) hash format
- Django(PBKDF2-HMAC-SHA256) hashcat mode added

v2.8.5
- added Cisco Type 8 & 9 hash format
- added Android FDE ≤ 4.3 hash format
- added Microsoft Office ≤ 2003 hash format
- added Microsoft Office 2007 hash format
- added Microsoft Office 2010 hash format
- added Microsoft Office 2013 hash format
- RAdmin v2.x hashcat mode added

v2.8.4
- added Lotus Notes/Domino 8 hash format
- added scrypt hash format
- added support for raw passwd & shadow files without preprocessing
- renamed Oracle 11g to Oracle 11g/12c
- renamed Samsung Android Password/PIN to Android PIN

v2.8.3
- removed unnecessary escaping of literal characters in regexes
- reworked man page and added license notice
- fixed some length bugs mainly in the base64 hash formats
- added Skype hash format

v2.8.2
- massive speed improvements
- added man page for hashID
- added PBKDF2(Atlassian) hash format
- added PostgreSQL MD5 hash format

v2.8.1
- added PeopleSoft hash format
- added Django(DES Crypt Wrapper) hash format
- added Django(PBKDF2-HMAC-SHA256) hash format
- added Django(PBKDF2-HMAC-SHA1) hash format
- added Django(BCrypt) hash format
- added Django(MD5) hash format
- Woltlab Burning Board 3.x regex fixes
- renamed MaNGOS CMS to MangosWeb Enhanced CMS
- renamed bcrypt to BCrypt
- renamed Django CMS(SHA-1) to Django(SHA-1)
- renamed Django CMS(SHA-256) to Django(SHA-256)
- renamed Django CMS(SHA-384) to Django(SHA-384)
- MangosWeb Enhanced CMS flagged as extended
- RAdmin v2.x flagged as extended
- ZipMonster flagged as extended
- Ventrilo flagged as extended

v2.8.0
- added new parameter "-a / --all"
- PEP8 optimization
- renamed xAuth to Minecraft(xAuth)

v2.7.0
- major rewrite
- dropped obsolete "-f / --file" and "-d / --dir" parameter
- all output defaults to stdout now

v2.6.7
- small cosmetic changes
- MediaWiki hashcat mode added

v2.6.6
- added 1Password(Agile Keychain) hash format
- added 1Password(Cloud Keychain) hash format
- added IKE-PSK MD5 hash format
- added IKE-PSK SHA1 hash format
- 1Password(Agile Keychain) hashcat mode added
- 1Password(Cloud Keychain) hashcat mode added
- IKE-PSK MD5 hashcat mode added
- IKE-PSK SHA1 hashcat mode added

v2.6.5
- hashID is usable as a python module now
- PHPS regex changes
- Kerberos 5 AS-REQ Pre-Auth regex changes
    
v2.6.4
- added PHPS hash format
- PHPS hashcat mode added

v2.6.3
- added Lotus Notes/Domino 5 hash format
- DNSSEC(NSEC3) regex fixes
- renamed Lotus Domino to Lotus Notes/Domino 6
- Lotus Notes/Domino 5 hashcat mode added
- Lotus Notes/Domino 6 hashcat mode added

v2.6.2
- added JTR formats to hashinfo.xlsx
- Juniper Netscreen/SSG(ScreenOS) regex changes
- NTLM regex changes

v2.6.1
- additional mimetype check on file input
- added MSSQL(2014) hash format
- renamed Blowfish(Eggdrop) to Eggdrop IRC Bot

v2.6.0
- added directory analyze using parameter "-d / --dir"
- more error checking on user input
- vBulletin < v3.8.5 regex fixes

v2.5.0
- added CRC-24 hash format
- added Joomla ≥ 2.5.18 hash format
- added Woltlab Burning Board 4.x hash format
- GRUB 2 regex changes
- BCrypt(SHA-256) regex changes
- OSX v10.8 regex changes
- OSX v10.9 regex changes
- DNSSEC(NSEC3) regex changes
- Minecraft(AuthMe Reloaded) regex changes
- renamed Wordpress 2.6.2 to Wordpress v2.6.2
- renamed Joomla to Joomla < 2.5.18
- renamed Drupal7 to Drupal ≥ v7.x
- renamed  Burning Board 3.x to Woltlab Burning Board 3.x
- hMailServer hashcat mode added

v2.4.5
- added Fairly Secure Hashed Password hash format
- SAP CODVN B (BCODE) regex changes
- SAP CODVN F/G (PASSCODE) regex changes
- renamed Cisco-IOS(SHA256) to Cisco-IOS(SHA-256)
    
v2.4.4
- added xAuth hash format
- added PBKDF2(Generic) hash format
- added PBKDF2(Cryptacular) hash format
- added PBKDF2(Dwayne Litzenberger) hash format

v2.4.3
-  added hMailServer hash format
-  added MediaWiki hash format
-  vBulletin ≥ v3.8.5 regex changes
-  Cisco-ASA(MD5) regex changes
-  renamed BCrypt(SHA256) to BCrypt(SHA-256)
    
v2.4.2
- added SHA-1 Crypt hash format
- renamed MD5(Sun) to Sun MD5 Crypt
- renamed SHA-256(Unix) to SHA-256 Crypt
- renamed SHA-512(Unix) to SHA-512 Crypt
- renamed MD5(Unix) to MD5 Crypt

v2.4.1
- added NTHash(FreeBSD Variant) hash format
- Domain Cached Credentials regex changes
- Domain Cached Credentials 2 regex changes
- NetNTLMv1-VANILLA / NetNTLMv1+ESS regex changes
- NetNTLMv2 regex changes

v2.4.0
- added Hashcat Mode output using parameter "-hc / --hashcat"
- added RACF hash format
- added Double MD5 hash format
- added Double SHA-1 hash format
- added md5apr1 hash format
- added bcrypt hash format
- NetNTLMv1-VANILLA / NetNTLMv1+ESS regex changes
- NetNTLMv2 regex changes
    
v2.3.6
- added Burning Board 3.x hash format
- added IPMI2 RAKP HMAC-SHA1 hash format
- added Lastpass hash format
- added Cisco-ASA(MD5) hash format
- added DNSSEC(NSEC3) hash format
- VNC regex changes
- renamed Keccak to SHA3

v2.3.5
- added Wordpress v2.6.0/2.6.1 hash format
- MyBB ≥ v1.2+ regex changes
- renamed MD5(phpBB) to phpBB 3.x
- renamed MD5(Wordpress) to Wordpress ≥ v2.6.2
- renamed MD5(ZipMonster) to ZipMonster
- renamed MD5(MaNGOS) to MaNGOS CMS
- renamed MD5(MaNGOS2) to MaNGOS CMS v2
- renamed SHA1(LinkedIn)to LinkedIn
- renamed SHA-1(Django) to Django CMS(SHA-1)
- renamed SHA-256(Django) to Django CMS(SHA-256)
- renamed SHA-384(Django) to Django CMS(SHA-384)

hashid.php
<?php
/******************************************************************************
* @name: hashIdentify.php
* @author: c0re <https://psypanda.org/>							
* @date: 2013/01/29
* @copyright: <http://creativecommons.org/licenses/by-nc-sa/3.0/>
* @readme: <http://wiki.insidepro.com/index.php/Algorithms>
******************************************************************************/
$version = "v0.7b";

?>

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <title>hashIdentifier</title>
</head>
<body>

<h1>hashIdentifier</h1>
<h4>[analyze your hash]</h4>

<!-- begin form -->
<form action="<?php echo htmlspecialchars($_SERVER['PHP_SELF']); ?>" method="post">
  <input type="text" name="hash" size="100" />
  <input type="submit" name="submit" value="submit" />
</form>
<!-- end form -->

<?php

function IdentifyHash($str)
{
  //initialize the array
  $possibleHashes = array();
	
  if(preg_match('/^[a-f0-9]{4}$/i', $str)) {
    array_push($possibleHashes,'CRC-16','CRC-16-CCITT','FCS-16');
  }
  if(preg_match('/^[a-f0-9]{8}$/i', $str)) {
    array_push($possibleHashes,'Adler32','CRC-32','CRC-32B','FCS-32','GHash-32-3','GHash-32-5');
  }
  if(preg_match('/^\+[a-z0-9\/\.]{12}$/i', $str)) {
    array_push($possibleHashes,'Blowfish(Eggdrop)');
  }
  if(preg_match('/^.{0,2}[a-z0-9\/\.]{11}$/i', $str)) {
    array_push($possibleHashes,'DES(Unix)');
  }
  if(preg_match('/^[a-f0-9]{16}$/i', $str)) {
    array_push($possibleHashes,'MySQL3.x','LM','DES(Oracle)','VNC');
  }
  if(preg_match('/^[a-z0-9\/\.]{16}$/i', $str)) {
    array_push($possibleHashes,'MD5(Cisco PIX)');
  }
  if(preg_match('/^\$1\$.{0,8}\$[a-z0-9\/\.]{22}$/i', $str)) {
    array_push($possibleHashes,'MD5(Unix)');
  }
  if(preg_match('/^\$apr1\$.{0,8}\$[a-z0-9\/\.]{22}$/i', $str)) {
    array_push($possibleHashes,'MD5(APR)');
  }
  if(preg_match('/^[a-f0-9]{24}$/i', $str)) {
    array_push($possibleHashes,'CRC-96(ZIP)');
  }
  if(preg_match('/^\$H\$9[a-z0-9\/\.]{30}$/i', $str)) {
    array_push($possibleHashes,'MD5(phpBB3)');
  }
  if(preg_match('/^\$P\$[a-z0-9\/\.]{31}$/i', $str)) {
    array_push($possibleHashes,'MD5(Wordpress)');
  }
  if(preg_match('/^[0-9a-f]{32}$/i', $str)) {
    array_push($possibleHashes,'MD5','NTLM','Domain Cached Credentials','Domain Cached Credentials 2','MD4','MD2','RIPEMD-128','Haval-128','Tiger-128','Snefru-128','Skein-256(128)','Skein-512(128)');
  }
  if(preg_match('/^0x[a-f0-9]{32}$/i', $str)) {
    array_push($possibleHashes,'Lineage II C4');
  }
  if(preg_match('/^[a-f0-9]{32}:[a-z0-9]{32}$/i', $str)) {
    array_push($possibleHashes,'MD5(Joomla)');
  }
  if(preg_match('/^[a-f0-9]{32}:.{5}$/i', $str)) {
    array_push($possibleHashes,'MD5(IP.Board)');
  }
  if(preg_match('/^[a-f-0-9]{32}:[a-z0-9]{8}$/i', $str)) {
    array_push($possibleHashes,'MD5(MyBB)');
  }
  if(preg_match('/^[a-f0-9]{40}$/i', $str)) {
    array_push($possibleHashes,'SHA-1','MySQL4.x','RIPEMD-160','Haval-160','SHA-1(MaNGOS)','SHA-1(MaNGOS2)','Tiger-160','Skein-256(160)','Skein-512(160)');
  }
  if(preg_match('/^\*[a-f0-9]{40}$/i', $str)) {
    array_push($possibleHashes,'MySQL5.x');
  }
  if(preg_match('/^sha1\$.{0,32}\$[a-f0-9]{40}$/i', $str)) {
    array_push($possibleHashes,'SHA-1(Django)');
  }
  if(preg_match('/^0x0100[a-f0-9]{0,8}?[a-f0-9]{40}$/i', $str)) {
    array_push($possibleHashes,'MSSQL(2005)','MSSQL(2008)');
  }
  if(preg_match('/^[a-f0-9]{48}$/i', $str)) {
    array_push($possibleHashes,'Haval-192','Tiger-192');
  }
  if(preg_match('/^[a-f0-9]{51}$/i', $str)) {
    array_push($possibleHashes,'MD5(Palshop)');
  }
  if(preg_match('/^\$S\$[a-z0-9\/\.]{52}$/i', $str)) {
    array_push($possibleHashes,'SHA-512(Drupal)');
  }
  if(preg_match('/^\$2a\$[0-9]{0,2}?\$[a-z0-9\/\.]{53}$/i', $str)) {
    array_push($possibleHashes,'Blowfish(OpenBSD)');
  }
  if(preg_match('/^[a-f0-9]{56}$/i', $str)) {
    array_push($possibleHashes,'SHA-224','Haval-224','Keccak-224','Skein-256(224)','Skein-512(224)');
  }
  if(preg_match('/^[a-f0-9]{64}$/i', $str)) {
    array_push($possibleHashes,'SHA-256','RIPEMD-256','Haval-256','Snefru-256','GOST R 34.11-94','Keccak-256','Skein-256','Skein-512(256)');
  }
  if(preg_match('/^sha256\$.{0,32}\$[a-f0-9]{64}$/i', $str)) {
    array_push($possibleHashes,'SHA-256(Django)');
  }
  if(preg_match('/^\$5\$.{0,22}\$[a-z0-9\/\.]{43,69}$/i', $str)) {
    array_push($possibleHashes,'SHA-256(Unix)');
  }
  if(preg_match('/^[a-f0-9]{80}$/i', $str)) {
    array_push($possibleHashes,'RIPEMD-320');
  }
  if(preg_match('/^0x0100[a-f0-9]{0,8}?[a-f0-9]{80}$/i', $str)) {
    array_push($possibleHashes,'MSSQL(2000)');
  }
  if(preg_match('/^\$6\$.{0,22}\$[a-z0-9\/\.]{86}$/i', $str)) {
    array_push($possibleHashes,'SHA-512(Unix)');
  }
  if(preg_match('/^[a-f0-9]{96}$/i', $str)) {
    array_push($possibleHashes,'SHA-384','Keccak-384','Skein-512(384)','Skein-1024(384)');
  }
  if(preg_match('/^sha384\$.{0,32}\$[a-f0-9]{96}$/i', $str)) {
    array_push($possibleHashes,'SHA-384(Django)');
  }
  if(preg_match('/^[a-f0-9]{128}$/i', $str)) {
    array_push($possibleHashes,'SHA-512','Whirlpool','Keccak-512','Skein-512','Skein-1024(512)');
  }
  if(preg_match('/^[a-f0-9]{256}$/i', $str)) {
    array_push($possibleHashes,'Skein-1024');
  }
	
  //no hash found
  if (empty($possibleHashes)) {
    return $possibleHashes = array('Unknow Hash');
  }
  //return the array
  else {
    return $possibleHashes;
  }
}

//check if the form got submitted
if(isset($_POST['submit']))
{
  //save it in a variable
  $hash = $_POST['hash'];
  
  //trim possible whitespaces
  $hash = trim($hash);
	
  //check if empty form submitted
  if(empty($hash) && !is_numeric($hash)) {
    die('<script type="text/javascript">alert("no input detected!")</script>');
  }
	
  //save the array
  $hashes = IdentifyHash($hash);
	
  //show most and less possible results
  if(count($hashes) > 2) {
    //save most possible hashes
    $mostpossible = array_slice($hashes,0,2);
    //print most possible results
    echo "<br /><b>Most Possible:</b><br />".implode("<br />",$mostpossible)."<br />";
    //save less possible hashes
    $hashes = (array_slice($hashes,2,count($hashes)));
    //print less possible results
    echo "<br /><b>Less Possible:</b><br />".implode("<br />",$hashes)."<br /><br />";
  }
  //show absolut result
  else {
    //print the results to the screen
    echo "<br />".implode("<br />",$hashes)."<br /><br />";
  }
}
?>

</body>
</html>